### PR TITLE
[Midend] Add RFFT op in Extend DAP Pass

### PR DESCRIPTION
**Pull request by BrokenArrow1404**
Add RFFT op to calc FFT for the real number. Algorithm from [PocketFFT ](https://gitlab.mpcdf.mpg.de/mtr/pocketfft/-/blob/master/pocketfft.c?ref_type=heads)library.

### CODE REVIEW

**Review by taiqzheng — CHANGES_REQUESTED** [review]

File: examples/DAPDialect/CMakeLists.txt [hunk 1]
```diff
@@ -62,3 +62,10 @@ target_link_libraries(buddy-whisper-preprocess
   BuddyLibDAP
   mlir_c_runner_utils
 )
+
+add_executable(buddy-whisper-preprocess-rfft WhisperPreprocessRFFT.cpp)
+add_dependencies(buddy-whisper-preprocess-rfft buddy-opt)
+target_link_libraries(buddy-whisper-preprocess-rfft
+  BuddyLibDAP
+  mlir_c_runner_utils
+)
```
Comment (taiqzheng) [inline examples/DAPDialect/CMakeLists.txt:71]: Please add an empty line at the end.

**Review by taiqzheng — CHANGES_REQUESTED** [review]

File: examples/DAPDialect/CMakeLists.txt [hunk 2]
```diff
@@ -62,3 +62,10 @@ target_link_libraries(buddy-whisper-preprocess
   BuddyLibDAP
   mlir_c_runner_utils
 )
+
+add_executable(buddy-whisper-preprocess-rfft WhisperPreprocessRFFT.cpp)
```
Comment (taiqzheng) [inline examples/DAPDialect/CMakeLists.txt:66]: please change target `buddy-whisper-preprocess-rfft` to `buddy-rfft`, for it is a standalone operator.

File: frontend/Interfaces/buddy/DAP/DSP/WhisperPreprocess.h [hunk 3]
```diff
@@ -49,6 +52,12 @@ void whisperPreprocess(MemRef<double, 1> *inputRawSpeech,
   detail::_mlir_ciface_buddy_whisperPreprocess(outputFeatures, inputRawSpeech);
 }
 
+
+void RFFT(MemRef<double, 1> *inputRawSpeech) {
+  detail::_mlir_ciface_buddy_RFFT(inputRawSpeech);
+}
+
```
Comment (taiqzheng) [inline frontend/Interfaces/buddy/DAP/DSP/WhisperPreprocess.h:59]: please remove two empty lines here.

File: frontend/Interfaces/lib/DAP-extend.mlir [hunk 4]
```diff
@@ -2,3 +2,8 @@ func.func @buddy_whisperPreprocess(%in : memref<?xf64>) -> memref<1x80x3000xf32>
   %out = dap.whisper_preprocess %in : memref<?xf64> to memref<1x80x3000xf32>
   return %out : memref<1x80x3000xf32>
 }
+func.func @buddy_RFFT(%in : memref<?xf64>) -> () {
+  dap.rfft %in : memref<?xf64>
+  // return %in : memref<400xf64>
```
Comment (taiqzheng) [inline frontend/Interfaces/lib/DAP-extend.mlir:7]: please remove the comment(line 7).

File: examples/DAPDialect/WhisperPreprocessRFFT.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  for (int i = 0; i < testLength; ++i) {
+    inputAlign[i] = static_cast<double>(i);
+  }
+  intptr_t inputSizes[1] = {testLength};
+  MemRef<double, 1> inputMemRef(inputAlign, inputSizes);
+
+  printLogLabel();
+  std::cout << "Preprocessing RFFT" << std::endl;
+  const auto loadStart = std::chrono::high_resolution_clock::now();
+
```
Comment (taiqzheng) [inline examples/DAPDialect/WhisperPreprocessRFFT.cpp:63]: please remove the empty line before `dap::RFFT`

File: examples/DAPDialect/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Buddy Compiler RFFT Example
+
+## Introduction
+
+This example demonstrates using the RFFT operator in Buddy Compiler to compute the Fourier transform of real numbers, with the algorithm derived from the [PocketFFT library](https://gitlab.mpcdf.mpg.de/mtr/pocketfft/-/tree/master?ref_type=heads).
+
+## How to run
+
```
Comment (taiqzheng) [inline examples/DAPDialect/README.md:8]: Please remove line 8~ line 66, if `buddy-rfft` doesn't need python binding.

File: examples/DAPDialect/RFFT.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
+//===- WhisperPreprocessor.cpp --------------------------------------------===//
```
Comment (taiqzheng) [inline examples/DAPDialect/RFFT.cpp:1]: `//===- WhisperPreprocessor.cpp --------------------------------------------===//`change to
`//===- RFFT.cpp - Example of DAP RFFT -------------------------------------===//`

File: frontend/Interfaces/buddy/DAP/DSP/WhisperPreprocess.h [hunk 8]
```diff
@@ -49,6 +52,10 @@ void whisperPreprocess(MemRef<double, 1> *inputRawSpeech,
   detail::_mlir_ciface_buddy_whisperPreprocess(outputFeatures, inputRawSpeech);
 }
 
+
```
Comment (taiqzheng) [inline frontend/Interfaces/buddy/DAP/DSP/WhisperPreprocess.h:55]: line 55: `// Function for RFFT`

**Review by taiqzheng — APPROVED** [review]
LGTM. Thanks!

**Review by zhanghb97 — APPROVED** [review]
Thanks!
Please add more `rfft` operation tests in your next PR!

**Pull request merged**